File: parser/lexer.go
package parser

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/rs/zerolog/log"
)

// Token represents a parsed token in Logo
type Token struct {
	Type  TokenType
	Value string
}

// TokenType defines the type of tokens
type TokenType string

const (
	CommandToken   TokenType = "COMMAND"
	NumberToken    TokenType = "NUMBER"
	RepeatToken    TokenType = "REPEAT"
	OpenBracket    TokenType = "OPEN_BRACKET"
	CloseBracket   TokenType = "CLOSE_BRACKET"
	VariableToken  TokenType = "VARIABLE"
	ProcedureToken TokenType = "PROCEDURE"
	ToToken        TokenType = "TO"
	EndToken       TokenType = "END"
	MakeToken      TokenType = "MAKE"
	IfToken        TokenType = "IF"
	StringToken    TokenType = "STRING"
	OperatorToken  TokenType = "OPERATOR"
	CommentToken   TokenType = "COMMENT"
)

// Lexer breaks input into tokens
type Lexer struct {
	input  string
	tokens []Token
}

// NewLexer creates a new lexer
func NewLexer(input string) *Lexer {
	return &Lexer{
		input: strings.TrimSpace(input),
	}
}

// Tokenize breaks the input into tokens
func (l *Lexer) Tokenize() error {
	tokens := []Token{}
	// Use a more flexible tokenization method
	input := l.input
	input = strings.ReplaceAll(input, "[", " [ ")
	input = strings.ReplaceAll(input, "]", " ] ")
	words := strings.Fields(input)

	for i := 0; i < len(words); i++ {
		word := strings.ToLower(words[i])

		// Handle comments
		if strings.HasPrefix(word, ";") {
			// Skip rest of line
			for i < len(words) {
				i++
			}
			continue
		}

		switch word {
		// Movement commands
		case "forward", "fd":
			tokens = append(tokens, Token{Type: CommandToken, Value: "forward"})
		case "backward", "bk":
			tokens = append(tokens, Token{Type: CommandToken, Value: "backward"})
		case "left", "lt":
			tokens = append(tokens, Token{Type: CommandToken, Value: "left"})
		case "right", "rt":
			tokens = append(tokens, Token{Type: CommandToken, Value: "right"})
		case "setx":
			tokens = append(tokens, Token{Type: CommandToken, Value: "setx"})
		case "sety":
			tokens = append(tokens, Token{Type: CommandToken, Value: "sety"})
		case "setheading", "seth":
			tokens = append(tokens, Token{Type: CommandToken, Value: "setheading"})
		case "home":
			tokens = append(tokens, Token{Type: CommandToken, Value: "home"})

		// Pen commands
		case "penup", "pu":
			tokens = append(tokens, Token{Type: CommandToken, Value: "penup"})
		case "pendown", "pd":
			tokens = append(tokens, Token{Type: CommandToken, Value: "pendown"})
		case "setpencolor", "setpc":
			tokens = append(tokens, Token{Type: CommandToken, Value: "setpencolor"})
		case "setpensize", "setps":
			tokens = append(tokens, Token{Type: CommandToken, Value: "setpensize"})

		// Control structures
		case "repeat":
			tokens = append(tokens, Token{Type: RepeatToken, Value: "repeat"})
		case "to":
			tokens = append(tokens, Token{Type: ToToken, Value: "to"})
		case "end":
			tokens = append(tokens, Token{Type: EndToken, Value: "end"})
		case "if":
			tokens = append(tokens, Token{Type: IfToken, Value: "if"})
		case "make":
			tokens = append(tokens, Token{Type: MakeToken, Value: "make"})

		// Brackets and operators
		case "[":
			tokens = append(tokens, Token{Type: OpenBracket, Value: "["})
		case "]":
			tokens = append(tokens, Token{Type: CloseBracket, Value: "]"})
		case "+", "-", "*", "/", "<", ">", "=":
			tokens = append(tokens, Token{Type: OperatorToken, Value: word})

		default:
			// Check if it's a number
			if num, err := strconv.ParseFloat(word, 64); err == nil {
				tokens = append(tokens, Token{Type: NumberToken, Value: fmt.Sprintf("%f", num)})
				continue
			}

			// Check if it's a variable (starts with ":")
			if strings.HasPrefix(word, ":") {
				tokens = append(tokens, Token{Type: VariableToken, Value: word[1:]})
				continue
			}

			// Check if it's a string (starts with ")
			if strings.HasPrefix(word, "\"") {
				tokens = append(tokens, Token{Type: StringToken, Value: word[1:]})
				continue
			}

			// Assume it's a procedure name
			tokens = append(tokens, Token{Type: ProcedureToken, Value: word})
		}
	}

	l.tokens = tokens

	// Log the parsed tokens
	for _, token := range tokens {
		log.Debug().Msgf("phase=lex token: %s:%s", token.Type, token.Value)
	}

	return nil
}

// GetTokens returns the parsed tokens
func (l *Lexer) GetTokens() []Token {
	return l.tokens
}

File: src/parser/lexer.ts
/**
 * Lexer for the Logo language
 * Breaks input text into tokens for parsing
 */
import { commandMap } from '../spec';

export enum TokenType {
  COMMAND,
  NUMBER,
  STRING,
  VARIABLE,
  REPEAT,
  TO,
  END,
  OPEN_PARENTHESIS,
  CLOSE_PARENTHESIS,
  OPEN_BRACKET,
  CLOSE_BRACKET,
  OPERATOR, // New: for expressions like +, -, *, /
  PROCEDURE,
  MAKE,
  IF,
  COMMENT
}

export interface Token {
  type: TokenType;
  value: string;
}

export class Lexer {
  private input: string;
  private tokens: Token[] = [];

  constructor(input: string) {
    this.input = input.trim();
  }

  /**
   * Tokenize the input string into an array of tokens
   */
  tokenize(): void {
    // Split the input by whitespace and filter out empty strings
    const words = this.input
      .replace(/\[/g, ' [ ')
      .replace(/\]/g, ' ] ')
      .replace(/\(/g, ' ( ')
      .replace(/\)/g, ' ) ')
      .replace(/([+\-*/])/g, ' $1 ') // Add spaces around basic operators
      .replace(/([<>]=?|==|!=)/g, ' $1 ') // Add spaces around comparison operators
      .split(/\s+/)
      .filter(word => word.length > 0);

    // Process each word
    for (const word of words) {
      // Check if the word is a number
      if (/^-?\d+(\.\d+)?$/.test(word)) {
        this.tokens.push({ type: TokenType.NUMBER, value: word });
        continue;
      }

      // Check if the word is a string (starts with ")
      if (word.startsWith('"')) {
        this.tokens.push({ type: TokenType.STRING, value: word.substring(1) });
        continue;
      }

      // Check if the word is a variable (starts with :)
      if (word.startsWith(':')) {
        this.tokens.push({ type: TokenType.VARIABLE, value: word.substring(1) });
        continue;
      }

      // Check if the word is an operator
      if (['+', '-', '*', '/', '<', '>', '<=', '>=', '==', '!='].includes(word)) {
        this.tokens.push({ type: TokenType.OPERATOR, value: word });
        continue;
      }

      if (word === '(') {
        this.tokens.push({ type: TokenType.OPEN_PARENTHESIS, value: '(' });
        continue;
      }

      if (word === ')') {
        this.tokens.push({ type: TokenType.CLOSE_PARENTHESIS, value: ')' });
        continue;
      }
      // Check for brackets
      if (word === '[') {
        this.tokens.push({ type: TokenType.OPEN_BRACKET, value: '[' });
        continue;
      }

      if (word === ']') {
        this.tokens.push({ type: TokenType.CLOSE_BRACKET, value: ']' });
        continue;
      }

      // Handle reserved control keywords first
      const lowerWord = word.toLowerCase();

      // Check for reserved words for procedure handling
      if (lowerWord === 'to') {
        this.tokens.push({ type: TokenType.TO, value: 'TO' });
        continue;
      }
      
      if (lowerWord === 'end') {
        this.tokens.push({ type: TokenType.END, value: 'END' });
        continue;
      }

      // Check if it's a defined command in our spec
      if (commandMap.has(lowerWord)) {
        const commandSpec = commandMap.get(lowerWord)!;
        // Use the official name from the command spec
        this.tokens.push({ type: TokenType.COMMAND, value: commandSpec.name.toLowerCase() });
        continue;
      }

      this.tokens.push({ type: TokenType.PROCEDURE, value: lowerWord });
    }

    // Log the parsed tokens (using console.log for browser environment)
    this.tokens.forEach(token => {
      console.debug(`[Lexer] token: ${TokenType[token.type]}:${token.value}`);
    });
  }

  /**
   * Returns the array of tokens
   */
  getTokens(): Token[] {
    return this.tokens;
  }
}
